include/vinecopulib/bicop/implementation/class.ipp

```diff
@@ -657,19 +804,46 @@ Bicop::select(const Eigen::Matrix<double, Eigen::Dynamic, 2>& data,
   }
 }
 
-//! @brief Data manipulations for rotated families
-//!
-//! @param u \f$m \times 2\f$ matrix of data.
-//! @return The manipulated data.
-inline Eigen::Matrix<double, Eigen::Dynamic, 2>
-Bicop::cut_and_rotate(const Eigen::Matrix<double, Eigen::Dynamic, 2>& u) const
+//! adds an additional column if there's only one discrete variable;
+//! removes superfluous columns for continuous variables. 
+//! (continuous models only require two columns, discrete models always four)
+inline Eigen::MatrixXd
+Bicop::format_data(const Eigen::MatrixXd& u) const
 {
-  Eigen::Matrix<double, Eigen::Dynamic, 2> u_new(u.rows(), 2);
+  int n_disc = get_n_discrete();
+  if (n_disc == 0) {
+    return u.leftCols(2);
+  } else if ((get_n_discrete() != 1) | (u.cols() == 4)) {
+    return u;
+  }
+  Eigen::MatrixXd u_new(u.rows(), 4);
+  u_new.leftCols(2) = u.leftCols(2);
+  int disc_col = (var_types_[1] == "d");
+  int cont_col = 1 - disc_col;
+  u_new.col(2 + disc_col) = u.col(2);
+  u_new.col(2 + cont_col) = u.col(cont_col);
+  return u_new;
+}
 
+//! clisp the data to the interval [1e-10, 1 - 1e-10] for numerical stability.
+inline Eigen::MatrixXd
+Bicop::clip_data(const Eigen::MatrixXd& u) const
+{
+  auto clip = [] (const double& x) {
```

Comment:
So, I've seen the trim function written a few times, always followed by `tools_eigen::unaryExpr_or_nan`  already above, and now this clip function... Shouldn't we add something in tools_eigen?

Like
```
Eigen::VectorXd trim(const Eigen::VectorXd& x, const down = 0, const up = o) const {
  auto trim_one = [down, up] (const double& x) {
      return std::min(std::max(x, down), 1 - up);
  };
  return tools_eigen::unaryExpr_or_nan(x, trim_one);
}
```